[Filebeat][httpjson] httpjson chain calls

kush-elastic opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
- Cleanup
- Docs
-->
- Enhancement
## What does this PR do?

- This PR contains code for extending httpjson to run chain calls and collect information from multiple urls.
- This httpjson code enables users to perform chains in a single manner.

<!-- Mandatory
Explain here the changes you made on the PR. Please explain the WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
-->

## Why is it important?

- Current filebeat httpjson input doesn't support chain calls.
For Example:
Salesforce: To collect eventLogFile from salesforce, we first need to collect logfile ids from Salesforce REST API and then use those logfile ids to collect more information from Salesforce.
<!-- Mandatory
Explain here the WHY, or the rationale/motivation for the changes.
-->

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [X] My code follows the style guidelines of this project
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] I have made corresponding change to the default configuration files
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [X] Check with Salesforce REST API to collect EventLogFile.



<!-- Recommended
## How to test this PR locally
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

## Related issues
 - Relates #22590
<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Superseds #123
-->

## Use cases
 - Configure chain parameters with each chain in step parameter.
<!-- Recommended
Explain here the different behaviors that this PR introduces or modifies in this project, user roles, environment configuration, etc.

If you are familiar with Gherkin test scenarios, we recommend its usage: https://cucumber.io/docs/gherkin/reference/
-->



<!-- Optional
## Screenshots
Add here screenshots about how the project will be changed after the PR is applied. They could be related to web pages, terminal, etc, or any other image you consider important to be shared with the team.
-->



<!-- Recommended
## Logs
Paste here output logs discovered while creating this PR, such as stack traces or integration logs, or any other output you consider important to be shared with the team.
-->

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-29816/detail/PR-29816/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-29816/detail/PR-29816/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-29816/detail/PR-29816/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-29816/detail/PR-29816/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://beats_29816.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2022-01-12T13:24:37.690Z&rangeTo=2022-01-12T13:44:37.690Z&transactionName=Beats%2Fbeats%2FPR-%7Bnumber%7D&transactionType=unknown&latencyAggregationType=avg&traceId=81d4fa415d8000ae845336c1ba332db6&transactionId=b82f31968cb01950)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-01-12T13:34:37.690+0000


* Duration: 15 min 53 sec


* Commit: 27e71a27c440ace7fd506e512bcb2174781d7f50



#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 3  |
| Skipped      | 0 |
| Total        | 3   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->











### Steps errors [![2](https://img.shields.io/badge/2%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-29816/detail/PR-29816/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `x-pack/filebeat-lint - make -C x-pack/filebeat check;make -C x-pack/filebeat update;make -C filebe`
<ul>
<li>Took 2 min 4 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-29816/runs/1/steps/1160/log/?start=0">here</a></li>
<li>Description: <code>make -C x-pack/filebeat check;make -C x-pack/filebeat update;make -C filebeat check;make -C filebeat update;make check-no-changes;</code></l1>
</ul>
  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-29816/runs/1/steps/1163/log/?start=0">here</a></li>
<li>Description: <code>Error 'hudson.AbortException: script returned exit code 2'</code></l1>
</ul>
  
  </p>
  </details>


## :green_heart: Flaky test report
Tests succeeded.



## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate the packages and run the E2E tests.


- `/beats-tester` : Run the installation tests with beats-tester.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

elasticmachine commented:
Pinging @elastic/integrations (Team:Integrations)

elasticmachine commented:
Pinging @elastic/security-external-integrations (Team:Security-External Integrations)

elasticmachine commented:
Pinging @elastic/elastic-agent-data-plane  (Team:Elastic-Agent-Data-Plane)

belimawr commented:
I'm reviewing it, I should have the review done tomorrow.

efd6 commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -146,38 +167,177 @@ func (rf *requestFactory) newHTTPRequest(stdCtx context.Context, trCtx *transfor
+			// perform request over collected ids
+			for _, id := range ids {
+				// reformat urls of requestFactory using ids
+				rf.url, err = generateNewUrl(rf.replace, urlString, id)
```
Is it expected that URL components outside the path will be altered? If not, `generateNewUrl` doesn't need to work on the entire URL string, just on the path as part of the `url.URL`. This will reduce allocs for rerendering.

kush-elastic replied:
For now, I am updating PR with other comments changes. and continue working on this comment.
based on your suggestion I am seeing some more possibilities like storing paths if the user wants or storing values in between steps.

kush-elastic replied:
Hey, I was looking into this one and decided to work with the entire URL instead of just the path. we are not sure but there can be a scenario where the user will be passing an entirely different URL in chain calls so, you know if we restrict users with just passing path we are reducing functionality there.

efd6 commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -88,24 +89,44 @@ type requestFactory struct {
 	password   string
 	log        *logp.Logger
 	encoder    encoderFunc
+	replace    string
```
Can this be compiled to a `*regexp.Regexp` during construction rather than repeatedly in `generateNewUrl`?

kush-elastic replied:
Right.

efd6 commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -146,38 +167,177 @@ func (rf *requestFactory) newHTTPRequest(stdCtx context.Context, trCtx *transfor
+					return err
+				}
+				r.responseProcessor.split = split
+				continue
+			} else {
```
`continue` here means that the following `else` block can be removed and the code dedented.

kush-elastic replied:
Right, we don't need `continue` there will remove it.

efd6 commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -146,38 +167,177 @@ func (rf *requestFactory) newHTTPRequest(stdCtx context.Context, trCtx *transfor
+		}
 	}
+
 	defer httpResp.Body.Close()
```
This doesn't need to be deferred since it's a final act anyway.

kush-elastic replied:
Right

efd6 commented on x-pack/filebeat/input/httpjson/response.go:
```diff
@@ -74,63 +74,68 @@ func newResponseProcessor(config *responseConfig, pagination *pagination, log *l
+						case errEmptyField:
+							// nothing else to send for this page
+							rp.log.Debug("split operation finished")
+							continue
```
`continue` is not necessary here and below.

kush-elastic replied:
Right.

efd6 commented on x-pack/filebeat/input/httpjson/response.go:
```diff
@@ -74,63 +74,68 @@ func newResponseProcessor(config *responseConfig, pagination *pagination, log *l
-						return
+					}
+
+					if err := rp.split.run(trCtx, tr, ch); err != nil {
+						switch err {
+						case errEmptyField:
```
```suggestion
					switch err := rp.split.run(trCtx, tr, ch); err {
					case nil:
					case errEmptyField:
```
(requires brace and white-space fixup below)

kush-elastic replied:
Got it. will update it.

belimawr commented on x-pack/filebeat/input/httpjson/json_parser.go:
```diff
@@ -0,0 +1,136 @@
+	"strings"
+)
+
+// parse returns array of string values from json string
```
Here you could add some more details of what parse does, maybe add some examples. Because it's the entrypoint function to this whole JSON parsing logic, it's good to have it well documented.

kush-elastic replied:
Sure. will add examples.

belimawr commented on x-pack/filebeat/input/httpjson/json_parser.go:
```diff
@@ -0,0 +1,136 @@
+
+// parse returns array of string values from json string
+func parse(rawJSON, key string) ([]string, error) {
+	var data interface{}
+	err := json.Unmarshal([]byte(rawJSON), &data)
+	if err != nil {
+		return nil, err
+	}
+	if err != nil {
+		return nil, fmt.Errorf("error while parsing json: %w", err)
+	}
+	values, err := parseInterface(data, key)
+	if err != nil {
+		return nil, fmt.Errorf("error while parsing json: %w", err)
+	}
+	return values, nil
```
You have duplicated error checks here, one function call and two consecutive blocks of `if err !=nil`. I also improved a bit the error message.
```suggestion
	var data interface{}
	if err := json.Unmarshal([]byte(rawJSON), &data); err != nil {
		return nil, fmt.Errorf("cannot unmarshal input data: %w", err)
	}

	values, err := parseInterface(data, key)
	if err != nil {
		return nil, fmt.Errorf("error while parsing JSON: %w", err)
	}

	return values, nil
```

kush-elastic replied:
sorry about that.

belimawr reviewed (changes requested):
Here is the first batch of review comments, so far I only looked at `x-pack/filebeat/input/httpjson/json_parser.go`. Some of my comments/suggestions (the ones I wrote yesterday evening) overlap with @efd6's ones. I kept them because they also have some variables renaming I believe increases readability.

I'll look into the other files now.

efd6 commented on x-pack/filebeat/input/httpjson/json_parser.go:
```diff
@@ -0,0 +1,136 @@
+// ]
+// output:
+// ["a_value_1", "a_value_2"]
+func jsonArr(key string, arrayInterface []interface{}) ([]interface{}, error) {
+	var arrayInterfaces []interface{}
+	for _, interfaces := range arrayInterface {
+		if _, ok := interfaces.(map[string]interface{}); ok {
+			if _, ok := interfaces.(map[string]interface{})[key]; ok {
+				arrayInterfaces = append(arrayInterfaces, interfaces.(map[string]interface{})[key])
+			} else {
+				return nil, fmt.Errorf("key field not found")
+			}
+		} else {
+			return nil, fmt.Errorf("invalid key")
+		}
+	}
+	return arrayInterfaces, nil
+}
```
With this, there is also the possibility of further reducing indent
```
	for _, el := range array {
		m, ok := el.(map[string]interface{})
		if !ok {
			return nil, fmt.Errorf("invalid key")
		}
		value, ok := m[key]
		if !ok {
			return nil, fmt.Errorf("key field not found")
		}
		result = append(result, value)
	}
```

kush-elastic replied:
Got it.

kush-elastic commented on x-pack/filebeat/input/httpjson/chain.go:
```diff
@@ -0,0 +1,16 @@
+
+package httpjson
+
+type chainConfig struct {
```
Got it. will add it.

efd6 replied:
I think some more detail would be appreciated by people that come here to fix bugs. Documentation on this type will likely be the first anchor that a new person will have on this code, so it should give them adequate context to be able to orient themselves.

Also, s/seperate/separate/.

kush-elastic commented on x-pack/filebeat/input/httpjson/json_parser.go:
```diff
@@ -0,0 +1,136 @@
+	for i, key := range splitKey {
+		switch key {
+		case "#":
+			// '#' means it's an array and we want to get the
+			// next key from each object in the array
+			if tmp, ok := data.([]interface{}); ok {
+				data, err = jsonArr(splitKey[i+1], tmp)
+				if err != nil {
+					return nil, fmt.Errorf("error while parsing json: %w", err)
+				}
+			}
+			// interate over []interface{}
+			if _, ok := data.([]interface{}); ok {
+				for _, jsoninterface := range data.([]interface{}) {
+					if _, ok := jsoninterface.(string); ok {
+						values = append(values, jsoninterface.(string))
+					} else {
+						joinKey := strings.Join(splitKey[i+2:], ".")
+						final, err := parseInterface(jsoninterface, joinKey)
+						if err != nil {
+							return nil, fmt.Errorf("error while parsing json: %w", err)
+						}
+						values = append(values, final...)
+					}
+				}
+			} else {
+				return nil, fmt.Errorf("error while parsing json: %w", err)
+			}
+		default:
+			// default assumes it's a JSON object, so it expects a
+			// map[string]interface{}
+			data, err = jsonNorm(key, data)
+			if err != nil {
+				return nil, fmt.Errorf("json is: %w", err)
+			}
+			if len(splitKey) == i+1 {
+				if _, ok := data.(string); ok {
+					values = append(values, data.(string))
+				}
+			}
```
Right. Thanks

kush-elastic commented on x-pack/filebeat/input/httpjson/json_parser.go:
```diff
@@ -0,0 +1,136 @@
+// 	"b": "b_value",
+// }
+func jsonNorm(key string, mapStr interface{}) (interface{}, error) {
+	if _, ok := mapStr.(map[string]interface{}); ok {
+		if _, ok := mapStr.(map[string]interface{})[key]; ok {
+			return mapStr.(map[string]interface{})[key], nil
```
Got it. will update it for the entire code.

kush-elastic commented on x-pack/filebeat/input/httpjson/json_parser.go:
```diff
@@ -0,0 +1,136 @@
+// map[string]interface{}{
+// 	"b": "b_value",
+// }
+func jsonNorm(key string, mapStr interface{}) (interface{}, error) {
+	if _, ok := mapStr.(map[string]interface{}); ok {
+		if _, ok := mapStr.(map[string]interface{})[key]; ok {
+			return mapStr.(map[string]interface{})[key], nil
+		}
+	}
+	return nil, fmt.Errorf("key field not found")
+}
```
Sure. will update it.

kush-elastic commented on x-pack/filebeat/input/httpjson/json_parser.go:
```diff
@@ -0,0 +1,136 @@
+	return arrayInterfaces, nil
+}
+
+// jsonNorm returns interface value from interface
+// input:
+// key=a
+// mapStr=map[string]interface{}{
+// 	"a": "a_value",
+// }
+// output:
+// a_value
+
+// input:
+// key=a
+// mapStr=map[string]interface{}{
+// 	"a": map[string]interface{}{
+// 		"b": "b_value",
+// 	},
+// }
+// map[string]interface{}{
+// 	"b": "b_value",
+// }
+func jsonNorm(key string, mapStr interface{}) (interface{}, error) {
```
Got it.

kush-elastic commented on x-pack/filebeat/input/httpjson/json_parser.go:
```diff
@@ -0,0 +1,136 @@
+	return values, nil
+}
+
+// jsonArr returns array of interface value from interface
+// input:
+// key=a,
+// arrayInterface=
+// [
+// 	{
+// 		"a": "a_value_1",
+// 	},
+// 	{
+// 		"a": "a_value_2",
+// 	},
+// ]
+// output:
+// ["a_value_1", "a_value_2"]
+func jsonArr(key string, arrayInterface []interface{}) ([]interface{}, error) {
```
Okay. will rename `jsonNorm` to `getArrayValue`. what do you think?
it will help others understand why it is different than getMapValue.

kush-elastic commented on x-pack/filebeat/input/httpjson/json_parser.go:
```diff
@@ -0,0 +1,136 @@
+			}
+			// interate over []interface{}
+			if _, ok := data.([]interface{}); ok {
+				for _, jsoninterface := range data.([]interface{}) {
```
Got it.

kush-elastic commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -146,38 +167,177 @@ func (rf *requestFactory) newHTTPRequest(stdCtx context.Context, trCtx *transfor
+				}
+				r.responseProcessor.split = split
+				continue
+			} else {
+				ids, err = r.getIdsFromResponses(intermediateResps, r.requestFactories[i+1].replace)
```
Right. Thanks

kush-elastic commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -206,8 +368,5 @@ func (r *requester) doRequest(stdCtx context.Context, trCtx *transformContext, p
 		n++
 	}
-
-	r.log.Infof("request finished: %d events published", n)
```
You can find it on line 308 now.

kush-elastic commented on x-pack/filebeat/input/httpjson/request_test.go:
```diff
@@ -133,3 +135,41 @@ func TestCtxAfterDoRequest(t *testing.T) {
+	type args struct {
+		replacement string
+		oldUrl      string
+		ids         string
```
I will be removing this test case. based on the next comment.

kush-elastic commented on x-pack/filebeat/input/httpjson/request_test.go:
```diff
@@ -133,3 +135,41 @@ func TestCtxAfterDoRequest(t *testing.T) {
+		ids         string
+	}
+	newurl, _ := url.Parse("http://some_url/id_10001/some_url")
+	tests := []struct {
+		name    string
+		args    args
+		want    url.URL
+		wantErr bool
+	}{
+		{
+			name: "will return old url with new id",
+			args: args{
+				replacement: "records.#.id",
+				oldUrl:      "http://some_url/records.#.id/some_url",
+				ids:         "id_10001",
+			},
+			want:    *newurl,
+			wantErr: false,
+		},
+	}
```
yeah, this test case is not required.

kush-elastic commented on x-pack/filebeat/input/httpjson/json_parser_test.go:
```diff
@@ -0,0 +1,432 @@
+		wantErr bool
+	}{
+		{
+			name: "will return string value from interface",
```
Okay.

kush-elastic commented on x-pack/filebeat/input/httpjson/json_parser_test.go:
```diff
@@ -0,0 +1,432 @@
+			wantErr: false,
+		},
+		{
+			name: "will return array of embedded interface values from interface",
```
Okay.

kush-elastic commented on x-pack/filebeat/input/httpjson/json_parser_test.go:
```diff
@@ -0,0 +1,432 @@
+			wantErr: false,
+		},
+		{
+			name: "can not collect values from non-existing key",
```
Okay.

kush-elastic commented on x-pack/filebeat/input/httpjson/json_parser_test.go:
```diff
@@ -0,0 +1,432 @@
+	}
+}
+
+func Test_parseInterface(t *testing.T) {
```
Right.

kush-elastic commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -146,38 +167,177 @@ func (rf *requestFactory) newHTTPRequest(stdCtx context.Context, trCtx *transfor
+	var err error
+	// collect ids from all responses
+	for _, resp := range intermediateResps {
+		if resp.Body != nil {
+			b, err = io.ReadAll(resp.Body)
+			if err != nil {
+				return nil, fmt.Errorf("error while reading response body: %w", err)
+			}
+		}
```
> But that makes me wonder if you really need access to the http.Response here. Maybe we could update requestFactory.collectResponse to return a []byte instead of the http.Response. That way the only bit of code that needs to worry about the quirks of HTTP responses (like closing the response body) is requestFactory.collectResponse. What do you think?
Right, but I am thinking like will it not cause problems for r.responseProcessor.startProcessing? it is using HTTP responses to process in pagination and all. so before closing it may cause problems for responseProcessor.

kush-elastic replied:
> Another thing I noticed is that there are no checks for the HTTP status code (or I missed them). The HTTP client we use does return an error if the status code is >= 400, that means response in the 1xx and 3xx ranges could reach this code path and will error because the data is not in the format we want or there is no data.
Is that by design? Or should we do some validation of the status code?

I think it is by design as I have noticed while collecting responses it is storing all types of responses even if it's not required.

efd6 commented on x-pack/filebeat/input/httpjson/json_parser.go:
```diff
@@ -0,0 +1,185 @@
+	splitKey := strings.Split(key, ".")
+	for i, key := range splitKey {
+		switch key {
+		case "#":
+			// '#' means it's an array and we want to get the
+			// next key from each object in the array
```
This appears to be new syntax for array indexing. Would it be worth considering something like JSONPath (or a subset of that) that already exists and has more power?

kush-elastic replied:
well, I will look into it.
I am not sure exactly but I think we can use jsonPATH.

kush-elastic replied:
Hey, I looked into it and found some existing libraries available for [JSONPath](https://pkg.go.dev/github.com/yalp/jsonpath).

As we discussed internally to not use any of such external libraries to process JSON, rather than to construct our own custom logic, and that's why we haven't used any of such libraries like [gjson](https://github.com/tidwall/gjson) or [JSONPath](https://pkg.go.dev/github.com/yalp/jsonpath).

@akshay-saraswat @mtojek

mtojek replied:
If this is just about a single function and we won't use 90% of any JSONPath library, then I would stick with a tiny custom implementation.

On the other hand, as we have @belimawr here, who is familiar with filebeat inputs, I would pass this question.

efd6 replied:
My main concern was inventing a new syntax.

belimawr replied:
https://github.com/elastic/beats/issues/22590 makes me believe this chain requests will be used in other places, so I see the possibility of this being used in other places.

If implementing (or adding a dependency) the full JSONPath is too much at the moment, we could use a subset of it, so later on if we need to extend this functionality we can just support the full JSONPath specification (thanks @cmacknz for this idea).

Anyway, I'm still reaching out to people so I can get a better high level view of this chain requests feature.

efd6 replied:
Putting this here in case it is helpful. There is an expression-extensible JSONPath [implementation for Go](https://github.com/PaesslerAG/jsonpath). This takes expressions using an [extensible language system](https://github.com/PaesslerAG/gval). [Example code](https://pkg.go.dev/github.com/PaesslerAG/jsonpath?utm_source=godoc#example-package--Gval).

efd6 commented on x-pack/filebeat/input/httpjson/json_parser.go:
```diff
@@ -0,0 +1,185 @@
+//   key=#.a
+// output:
+//   ["a_value_1", "a_value_2"]
+func parse(rawJSON, key string) ([]string, error) {
```
Make `rawJSON` a `[]byte` since the call site at L327 converts a `[]byte` to a string to satisfy this signature, resulting in an unnecessary allocation.

Suggest rename `rawJSON` to something like `data` and function that more clearly reflects its function; this is not a parser (same below). `getKeyedArrayValue`? to mirror the naming below.

kush-elastic replied:
Got it. will update it.

efd6 commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -146,38 +168,175 @@ func (rf *requestFactory) newHTTPRequest(stdCtx context.Context, trCtx *transfor
+			}
+			rf.url = urlCopy
+
+			if i == rfSize-1 {
+				split = r.responseProcessor.split
+				r.responseProcessor.split = rf.split
+				n, err = r.processAndPublishEvents(stdCtx, trCtx, publisher, finalResps, true)
+				if err != nil {
+					return err
+				}
+				r.responseProcessor.split = split
+			} else {
+				// The if comdition (i == rfSize-1) ensures this branch never runs to the last element
+				// of r.requestFactories, therefore r.requestFactories[i+1] will never be out of bounds.
+				ids, err = r.getIdsFromResponses(intermediateResps, r.requestFactories[i+1].replace)
+				if err != nil {
+					return err
+				}
+				split = r.responseProcessor.split
+				r.responseProcessor.split = rf.split
+				n, err = r.processAndPublishEvents(stdCtx, trCtx, publisher, intermediateResps, false)
+				if err != nil {
+					return err
+				}
+				r.responseProcessor.split = split
```
```suggestion
			var resps []*http.Response
			if i < rfSize {
				resps = finalResps
			} else {
				// The if comdition (i < rfSize) ensures this branch never runs to the last element
				// of r.requestFactories, therefore r.requestFactories[i+1] will never be out of bounds.
				resps = intermediateResps
				ids, err = r.getIdsFromResponses(resps, r.requestFactories[i+1].replace)
				if err != nil {
					return err
				}
			}
			split = r.responseProcessor.split
			r.responseProcessor.split = rf.split
			n, err = r.processAndPublishEvents(stdCtx, trCtx, publisher, resps, i < rfSize)
			if err != nil {
				return err
			}
			r.responseProcessor.split = split
```

Also (not done here), I'd recommend using `len(r.requestFactories)` in stead of `rfSize`. A length look-up is cheap and this keeps the semantics local. It's a matter of style though; people may disagree.

kush-elastic replied:
Okay

belimawr commented:
@kush-elastic I've seen you did quite some changes, have you finished addressing the comments? If you're done, could you use the "re-request review" button? That really helps to know when the PR is ready for a re-review.

kush-elastic commented:
> @kush-elastic I've seen you did quite some changes, have you finished addressing the comments? If you're done, could you use the "re-request review" button? That really helps to know when the PR is ready for a re-review.

Yes, I completed requested changes. please let me know if there are any remains. Thanks

efd6 commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -146,38 +168,170 @@ func (rf *requestFactory) newHTTPRequest(stdCtx context.Context, trCtx *transfor
+	for _, resp := range intermediateResps {
+		resp := resp
+		if resp.Body != nil {
+			defer resp.Body.Close()
```
This can be placed immediately after the call to `io.ReadAll` (before the error check) without the `defer`.

kush-elastic replied:
Got it.

efd6 commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -146,38 +168,170 @@ func (rf *requestFactory) newHTTPRequest(stdCtx context.Context, trCtx *transfor
+				if err != nil {
+					return err
+				}
+				continue
+			} else {
```
The else block is not required since in the `len(r.requestFactories) == 1` case you hit the `continue`, so remove it and de-indent the code in the body.

kush-elastic replied:
Right

belimawr commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -146,38 +168,169 @@ func (rf *requestFactory) newHTTPRequest(stdCtx context.Context, trCtx *transfor
+	for _, resp := range intermediateResps {
+		resp := resp
+		if resp.Body != nil {
+			b, err = io.ReadAll(resp.Body)
+			resp.Body.Close()
```
There are two tings worrying me here:
### 1. Read timeouts
Here we read the body of a request that was made on [L227](https://github.com/elastic/beats/blob/3aa2a1148592e44ca41b6c1dc81af9e0624182b6/x-pack/filebeat/input/httpjson/request.go#L227) or [L226](https://github.com/elastic/beats/blob/3aa2a1148592e44ca41b6c1dc81af9e0624182b6/x-pack/filebeat/input/httpjson/request.go#L264). The problem I see is the `for` loop on [L256](https://github.com/elastic/beats/blob/3aa2a1148592e44ca41b6c1dc81af9e0624182b6/x-pack/filebeat/input/httpjson/request.go#L256), if that loop takes too long to execute (the requests are slow or there are too many of them), for each request that was made Filebeat will keep the socket open and won't read any data until we reach this bit of the code. If the waiting time until we start reading those responses is longer than the set timeout (on Filebeat or on the remote server) then this read will fail.

### 2. The read and close of the body are quite far from where the request is made
Every time we make an HTTP request we need to make sure we close the response body, but because there are other parts of the code that could error (like [L258-L261](https://github.com/elastic/beats/blob/3aa2a1148592e44ca41b6c1dc81af9e0624182b6/x-pack/filebeat/input/httpjson/request.go#L258-L261)) before we reach the point where we read and close the response body, we cannot guarantee that will always happen. That can become a big problem because Filebeat will start leaking open file descriptors (sockets). Also the external system we're calling could suffer from a similar issue if they don't timeout/cancel the connection properly.

## A possible solution
To solve both issues, I suggest to refactor `collectResponse` so it reads (and closes) the response body right after the request is made, then it can return a `[]byte` instead of an `http.Response`. By what I could see in the code you don't use any other field of the response(is that right?). So I believe it should be safe to implement something like what I suggested.

kush-elastic commented on x-pack/filebeat/input/httpjson/chain.go:
```diff
@@ -0,0 +1,135 @@
+//
+// example configuration:
+//
+// chain:
+//     - step:
+//         request.url: https://some_url.com/services/data/v1.0/records.#.Id/export_ids
+//         request.method: GET
+//         replace: records.#.Id
+//     - step:
+//         request.url: https://some_url.com/services/data/v1.0/export_ids/export.data.#.Id/info
+//         request.method: GET
+//         replace: export.data.#.Id
```
Oh right, it really was misleading.
what I was trying to explain here was something like this:

```suggestion
// - type: httpjson
//   enabled: true
//   request.url: https://some_url.com/services/data/v1.0/records (normal call)
//   interval: 1h
// chain:
//     - step:
//         request.url: https://some_url.com/services/data/v1.0/records.#.Id/export_ids (step-1)
//         request.method: GET
//         replace: records.#.Id
//     - step:
//         request.url: https://some_url.com/services/data/v1.0/export_ids/export.data.#.Id/info (step-2)
//         request.method: GET
//         replace: export.data.#.Id
```

Thanks for pointing me to it.
I will update it accordingly.

ruflin commented:
In https://github.com/elastic/beats/pull/29816#issuecomment-1026998542 I raised some general concerns around the direction this PR is going. We had an internal discussion around this and what I learned is that this is the path forward that was already agreed on. Sorry for the noise and please keep pushing it forward.

kush-crest commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -88,24 +89,45 @@ type requestFactory struct {
-	return rf
+	rfs = append(rfs, rf)
+	for _, ch := range config.Chain {
+		reg, _ := regexp.Compile(ch.Step.Replace)
```
Okay

kush-elastic replied:
Okay, will update it.

kush-crest commented on x-pack/filebeat/input/httpjson/chain.go:
```diff
@@ -0,0 +1,135 @@
+//
+// request-url-1: https://some_url.com/services/data/v1.0/export_ids/export_id-4/info
+//
+// Collect all the responses from newly constructed URLs.
```
Yes, we can. using JSONPath. in step configuration we are using replace parameter to get which data to collect using path mentioned from the user.

kush-elastic replied:
Yes, we can. using JSONPath. in step configuration we are using replace parameter to get which data to collect using path mentioned from the user.

efd6 commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -96,24 +100,44 @@ type requestFactory struct {
-		rf.user = authConfig.Basic.User
-		rf.password = authConfig.Basic.Password
+	rfs = append(rfs, rf)
+	for _, ch := range config.Chain {
```
This terminates after the first level of the nesting of the `Chain` field. The original design considered the full n-ary tree of chains which would require a recursive construction here rather than a flattening into `rfs`. The power gained by that may not be necessary (it would allow bounded conditional execution, which was the design motivation at the time), but if it is, that construction is needed. If not, there should be documentation that notes that deeper nesting is ignored.

kush-elastic replied:
Yes @efd6 , I remember this was one of the implementation proposals that was discussed during our design meeting. We discussed it with the team and to make the implementation simpler, we thought of using the current approach. And I agree with your point, we should definitely call this out in the documentation that nesting is not allowed to avoid any confusion.

kush-elastic commented on x-pack/filebeat/input/httpjson/chain.go:
```diff
@@ -0,0 +1,146 @@
+	Step stepConfig `config:"step" validate:"required"`
+}
+
+// Request call can be configured in step configuration.
+// step configuration will contain basic properties like, request.url,
+// request.method and replace parameter. Each step: request.url
+// will contain replace string with original URL to make a skeleton for the
+// call request.
+type stepConfig struct {
```
Sure, I will update it.
Thanks

kush-elastic commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -154,38 +178,180 @@ func (rf *requestFactory) newHTTPRequest(stdCtx context.Context, trCtx *transfor
+		case map[string]interface{}:
+			ids = append(ids, fmt.Sprintf("%v", tresp))
+		default:
+			r.log.Debugf("not able to collect ")
```
Right, will update it.

kush-elastic commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -154,38 +178,180 @@ func (rf *requestFactory) newHTTPRequest(stdCtx context.Context, trCtx *transfor
+				ids = append(ids, fmt.Sprintf("%v", value))
+			}
+		case map[string]interface{}:
+			ids = append(ids, fmt.Sprintf("%v", tresp))
```
Oh, I have made a mistake there. what I wanted to do is if the user has just one value to collect then it will be in `int` or `string` or `double` format so I can directly append those to ids. if if the type of interface is a `map[string]interface{}` then I will show an error that we can not collect ids from it.

kush-elastic replied:
I will update it accordingly. Thanks

andrewkroh commented on x-pack/filebeat/docs/inputs/input-httpjson.asciidoc:
```diff
@@ -931,6 +932,149 @@ This will output:
 
+[[chain]]
+[float]
+=== `chain`
```
The docs contain a sequence diagram that explains the high level flow. I think the chain feature should be added to the diagram.

belimawr commented on x-pack/filebeat/input/httpjson/config.go:
```diff
@@ -17,6 +17,7 @@ type config struct {
 	Request  *requestConfig  `config:"request" validate:"required"`
 	Response *responseConfig `config:"response"`
 	Cursor   cursorConfig    `config:"cursor"`
+	Chain    []chainConfig   `config:"chain"`
```
[Question]
`chainConfig` only contains a single field `stepConfig`, is it really needed? Are they both used so the final yaml looks like that:
```yaml
   chain:
     - step:
         request.url: https://some_url.com/services/data/v1.0/$.records[:].id/export_ids (second call)
         request.method: GET
         replace: $.records[:].id
     - step:
         request.url: https://some_url.com/services/data/v1.0/export_ids/$.file_name/info (third call)
         request.method: GET
         replace: $.file_name
```

Maybe we could use the `stepConfig` directly:
```suggestion
	Chain    []stepConfig   `config:"chain"`
```
I believe it would result in a yaml like this:
```yaml
chain:
  - request.url: 'https://some_url.com/services/data/v1.0/$.records[:].id/export_ids'
    request.method: GET
    replace: '$.records[:].id'
  - request.url: 'https://some_url.com/services/data/v1.0/export_ids/$.file_name/info'
    request.method: GET
    replace: $.file_name

```

efd6 replied:
Yes, this is related to https://github.com/elastic/beats/pull/29816#discussion_r815601148. Changing type would be better than just documenting the situation, though does this preclude extending the computation model in the future, and if it does, does that matter.

kush-elastic replied:
Right, we thought of that but then decided to go with using a step to make it easy in the future to extend it to add some more processing in-between steps whenever required, for example:
```- type: httpjson
  enabled: true
  request.url: https://example.com/
  interval: 1h
  chain:
	- step:
		request.method: GET
		request.url: "http://example.com/auth"
		response: ...

	- break:
		when: '[[eq .last_response.body.authtok ""]]'

	- step:
		request.method: POST
		request.url: 'http://example.com/api?auth=[[.last_response.body.authtok]]'
		response: ...```
Hence, we were planning to update the docs for the user to mention using steps.

belimawr commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -154,38 +178,185 @@ func (rf *requestFactory) newHTTPRequest(stdCtx context.Context, trCtx *transfor
+			for _, v := range tresp {
+				_, ok := v.(map[string]interface{})
+				if ok {
+					r.log.Debugf("events must be int, string or double, but got %T: skipping", v)
+					continue
+				}
+				ids = append(ids, fmt.Sprintf("%v", v))
+			}
+		case map[string]interface{}:
+			r.log.Debugf("cannot collect IDs from type '%T' : '%v'", values, values)
```
Those logs are showing errors that will affect the final result, they should be at least level warning, maybe even error
```suggestion
					r.log.Errorf("events must be int, string or double, but got %T: skipping", v)
					continue
				}
				ids = append(ids, fmt.Sprintf("%v", v))
			}
		case map[string]interface{}:
			r.log.Errorf("cannot collect IDs from type '%T' : '%v'", values, values)
```

kush-elastic replied:
Got it.

belimawr commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -154,38 +178,185 @@ func (rf *requestFactory) newHTTPRequest(stdCtx context.Context, trCtx *transfor
+
+		// get replace values from collected json
+		var v interface{}
+		json.Unmarshal(b, &v)
```
I believe this unmarshal could fail (e.g: an invalid JSON), it's better to check and handle the error just to be on the safe side
```suggestion
		if err:= json.Unmarshal(b, &v); err !=nil{
		    return nil, fmt.Errorf("cannot unmarshal data: %w", err)
		}
```

kush-elastic replied:
Got it.

belimawr commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -154,38 +178,185 @@ func (rf *requestFactory) newHTTPRequest(stdCtx context.Context, trCtx *transfor
+			}
+		case map[string]interface{}:
+			r.log.Debugf("cannot collect IDs from type '%T' : '%v'", values, values)
+		default:
+			ids = append(ids, fmt.Sprintf("%v", tresp))
+		}
```
It's not safe to assume all other cases will work. Instead you can write another case with the accepted values and use a generic `default` for the error message:

```suggestion
		case float64, string:
			ids = append(ids, fmt.Sprintf("%v", tresp))
		default:
			r.log.Errorf("cannot collect IDs from type '%T': '%v'", values, values)
		}
```

belimawr reviewed (changes requested):
Great Job! A couple small details about logging/error handling.

belimawr commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -154,38 +178,187 @@ func (rf *requestFactory) newHTTPRequest(stdCtx context.Context, trCtx *transfor
+		switch tresp := values.(type) {
+		case []interface{}:
+			for _, v := range tresp {
+				_, ok := v.(map[string]interface{})
+				if ok {
+					r.log.Errorf("events must be int, string or double, but got %T: skipping", v)
+					continue
+				}
+				ids = append(ids, fmt.Sprintf("%v", v))
```
There are other unsupported types aside `map[string]interface{}`, e.g: another array. [This Go playground](https://go.dev/play/p/L31MR9vxb8r) contains an example where `v` could be an slice of slices.

To make the code safer, we should ensure `v` is from a supported type and if not, then handle the error.

```suggestion
				switch v.(type) {
				case float64, string:
					ids = append(ids, fmt.Sprintf("%v", v))
				default:
					r.log.Errorf("events must a number or string, but got %T: skipping", v)
					continue
				}
```

kush-elastic replied:
Right.

belimawr reviewed (changes requested):
I believe that's the last little detail @kush-elastic !

belimawr reviewed (changes requested):
Great job @kush-elastic !!

One last thing I've realised only now: there are no tests for this feature, could you add some? Sorry I didn't spot it before, I remember reviewing lots of test cases, but it seems they were from the old parser implementation we replaced by jsonpath.

You can focus them on the higher level bits, going from a config to the final data and mocking the external server with [`httptest`](https://pkg.go.dev/net/http/httptest)

marc-gr commented on x-pack/filebeat/docs/inputs/input-httpjson.asciidoc:
```diff
@@ -931,6 +932,149 @@ This will output:
+[float]
+==== `chain[].step.replace`
+
+replace is JSONPath string to parse values from response JSONs, collected from previous chain calls. Please look at the official doc of https://goessner.net/articles/JsonPath/index.html#e2[JSONPath]. Required.
```
```suggestion
A [JSONPath](https://goessner.net/articles/JsonPath/index.html#e2[JSONPath]) string to parse values from responses JSON, collected from previous chain steps. Required.
```

Is this always required? Can there be situations were a step does not expect a replaceable part?

kush-elastic replied:
No, It's always required.

marc-gr commented on x-pack/filebeat/docs/inputs/input-httpjson.asciidoc:
```diff
@@ -931,6 +932,149 @@ This will output:
+
+Example:
+
+first call: https://example.com/services/data/v1.0/
+
+second call: https://example.com/services/data/v1.0/1/export_ids
+
+third call: https://example.com/services/data/v1.0/export_ids/file_1/info
```
```suggestion
- First call: https://example.com/services/data/v1.0/

- Second call: https://example.com/services/data/v1.0/1/export_ids

- Third call: https://example.com/services/data/v1.0/export_ids/file_1/info
```

kush-elastic replied:
Okay

marc-gr commented on x-pack/filebeat/docs/inputs/input-httpjson.asciidoc:
```diff
@@ -931,6 +932,149 @@ This will output:
+
+Example:
+
+- first call: https://example.com/services/data/v1.0/
++
+<<response-json1,response>>
+
+- second call: https://example.com/services/data/v1.0/`$.records[:].id`/export_ids
++
+<<response-json2,response>>
+
+third call: https://example.com/services/data/v1.0/export_ids`/`$.file_name`/info
```
```suggestion
- First call: https://example.com/services/data/v1.0/
+
<<response-json1,response>>

- Second call: https://example.com/services/data/v1.0/`$.records[:].id`/export_ids
+
<<response-json2,response>>

- Third call: https://example.com/services/data/v1.0/export_ids`/`$.file_name`/info
```

marc-gr commented on x-pack/filebeat/docs/inputs/input-httpjson.asciidoc:
```diff
@@ -931,6 +932,149 @@ This will output:
+Collect and make events from response in any format supported by httpjson for all calls.
+
+NOTE: httpjson chain will only create and ingest events from last call on chain configurations.
+httpjson chain will also 
```
I think this text is incomplete

kush-elastic replied:
Oh, Right.

marc-gr commented on x-pack/filebeat/input/httpjson/chain.go:
```diff
@@ -0,0 +1,74 @@
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+// Example:
```
Maybe would be good to have this examples in the docs if they are different from the current ones

marc-gr reviewed (commented):
Looking good. As @andrewkroh mentioned, would be nice to update the flow diagram to represent this new step and make the interaction clear. 

Also would be nice if @dedemorton could give a look at the docs formatting to be sure the feature is clear.

marc-gr commented:
/test

kush-elastic commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -197,15 +371,17 @@ func (r *requester) doRequest(stdCtx context.Context, trCtx *transformContext, p
+			if err := publisher.Publish(event, trCtx.cursorMap()); err != nil {
+				r.log.Errorf("error publishing event: %v", err)
+				continue
+			}
```
Here I am trying to use existing cursor even if response are from chain call, so let's say I want to store value from chain call and use it in main call next time I can do that using cursor. we actually implemented this in salesforce module.
here's [link](https://github.com/kush-elastic/beats/blob/salesforce_observability_connector/x-pack/filebeat/module/salesforce/login-rest/config/login-rest.yml#L33).

marc-gr replied:
I am worried this breaks current behavior. In the docs (https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-input-httpjson.html#input-state) it is mentioned that 

```
last_event: A map representing the last event sent to the output (result from applying transforms to last_response.body)
```

If we update last_event for every chain, regardless of the event being published, this statement is not true anymore, and since cursor can reference last_event, it opens room for potentially confusing situations. 

I see that for your use case, you are only using one chain step, which would make it to behave the same with my suggestion. Do you think is okay to preserve current behavior (and maybe update cursor docs to make clear that updates on published events, also), and maybe we can see how to enhance this if needed in the future?

kush-elastic replied:
Sorry, My mistake.
But we need this as we are collecting it from first response and using it again in first call.

kush-elastic replied:
Can't we update doc with like `last_event of specific call`.

kush-elastic commented on x-pack/filebeat/docs/inputs/input-httpjson.asciidoc:
```diff
@@ -931,6 +932,149 @@ This will output:
+[float]
+=== `chain`
+
+Chain is a list of calls to be made after the first call.
+
+chain list option: [`step`].
+
```
Got it.

kush-elastic commented on x-pack/filebeat/docs/inputs/input-httpjson.asciidoc:
```diff
@@ -931,6 +932,149 @@ This will output:
+[float]
+==== `chain[].step`
+
+chain[].step will contain basic request and response configurations for chain calls.
```
Got it.

kush-elastic commented on x-pack/filebeat/docs/inputs/input-httpjson.asciidoc:
```diff
@@ -931,6 +932,149 @@ This will output:
+[float]
+==== `chain[].step.respose.split`
+
+Please refer <<response-split,response split parameter>>. Required.
```
No, sorry.
My mistake.

kush-elastic commented on x-pack/filebeat/docs/inputs/input-httpjson.asciidoc:
```diff
@@ -931,6 +932,149 @@ This will output:
+third call: https://example.com/services/data/v1.0/export_ids/file_1/info
+
+[float]
+==== `chain[].step.respose.split`
```
Right.

kush-elastic commented on x-pack/filebeat/docs/inputs/input-httpjson.asciidoc:
```diff
@@ -931,6 +932,149 @@ This will output:
+[float]
+==== `chain[].step.request`
+
+Please refer <<request-parameters,request parameters>>. also place same replace string in url where collected values from previous call should be place. Required. 
```
Got it.

kush-elastic commented on x-pack/filebeat/docs/inputs/input-httpjson.asciidoc:
```diff
@@ -931,6 +932,149 @@ This will output:
+
+Please refer <<request-parameters,request parameters>>. also place same replace string in url where collected values from previous call should be place. Required. 
+
+Example:
```
Right, As value can be different. user can mentioned different calls in specified way.

kush-elastic commented:
Hey @marc-gr, 
can you help me with this?

> As @andrewkroh mentioned, would be nice to update the flow diagram to represent this new step and make the interaction clear.

It's in `png` if possible can you get me this original file and whichever tool have been used for creating it.

marc-gr commented on x-pack/filebeat/input/httpjson/input_test.go:
```diff
@@ -389,6 +390,76 @@ func TestInput(t *testing.T) {
 			expected: []string{`{"hello":[{"world":"moon"},{"space":[{"cake":"pumpkin"}]}]}`},
 		},
+		{
+			name: "Test simple Chain GET request",
```
would be nice to add non trivial tests too (multiple chains, use of cursor, etc) to ensure the whole new feature set works as expected

belimawr reviewed (commented):
Also could you look into the conflicts?

belimawr reviewed (approved):
Approving so it won't be blocked by me while I'm away.

@kush-elastic please address the comments from @marc-gr and fix the conflicts before merging.

kush-elastic commented on x-pack/filebeat/input/httpjson/input_test.go:
```diff
@@ -389,6 +390,76 @@ func TestInput(t *testing.T) {
+		{
+			name: "Test simple Chain GET request",
+			setupServer: func(t *testing.T, h http.HandlerFunc, config map[string]interface{}) {
+				r := mux.NewRouter()
+				r.HandleFunc("/", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+					fmt.Fprintln(w, `{"records":[{"id":1}]}`)
+				}))
+				r.HandleFunc("/1", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+					fmt.Fprintln(w, `{"hello":[{"world":"moon"},{"space":[{"cake":"pumpkin"}]}]}`)
+				}))
+				server := httptest.NewServer(r)
```
Thanks @belimawr,
I will update this.

kush-elastic commented:
Hey @marc-gr,
I can see few errors in this newly added golangci-lint. I do not see the line numbers or file names associated with these errors which is making it very difficult to isolate and fix them. For instance, for the Error: ` comparing with == will fail on wrapped errors. Use errors.Is to check for a specific error (errorlint)`  is searched over all the files modified as a part of this PR and none of them has this issue. Can you help us fix this so we can get this merged since all the comments are addressed and this is the only thing stopping us from merging this PR.

efd6 commented:
@kush-elastic I have a PR which will resolve the lint issues that are unrelated to the changes you have here. This should be merged soon.

kush-elastic commented on x-pack/filebeat/input/httpjson/input_test.go:
```diff
@@ -389,6 +392,195 @@ func TestInput(t *testing.T) {
+					},
+				},
+			},
+			handler:  defaultHandler("GET", ""),
```
It's not complaining.
should I update, now?

efd6 replied:
It must be below the threshold (3 instances). For consistency with the rest of the package and to avoid it tipping when another one is added, I think it should be changed (all instances).

efd6 replied:
Well, looking ... I'm wrong, there are 18 instance. I don't know why it's not complaining, but it should be changed anyway.

efd6 reviewed (approved):
LGTM, but please wait for others.

marc-gr commented:
/test

rdner commented:
@mtojek merging this PR might have broken the documentation, see https://elasticsearch-ci.elastic.co/blue/organizations/jenkins/elastic%2Bbeats%2Bpull-request%2Bbuild-docs/detail/elastic%20beats%20pull-request%20build-docs/13929/pipeline/

it's hard to tell the exact cause but at least I see these warnings

```
INFO:build_docs:asciidoctor: WARNING: ../../x-pack/filebeat/docs/inputs/input-httpjson.asciidoc: line 1042: id assigned to block already in use: response-json1
INFO:build_docs:asciidoctor: WARNING: invalid reference: response-json2
```

szabosteve commented:
@mtojek 

The first issue is caused by this line: https://github.com/elastic/beats/pull/29816/files#diff-531162d76f10990e57691d2298a46f2aa2a4e576a8ec0d17dd1a37c35303983fR1040 (as the ID is already in use).
The second one is caused by this: https://github.com/elastic/beats/pull/29816/files#diff-531162d76f10990e57691d2298a46f2aa2a4e576a8ec0d17dd1a37c35303983fR977 (response-json2 is an invalid reference)

ruflin commented:
@mtojek Why is this planned to be backported to 8.2? This is not a bug fix.

mtojek commented:
@ruflin isn't after the feature freeze now?

dedemorton commented:
Sorry, I missed the review because I was on PTO. :/ Let's not worry too much about the structure right now. We'll be moving to a new system and need to do some refactoring/cleanup as part of the migration, so let's handle layout issues then.
